### ISSUE TITLE
Increasing epoch length during iterative Resnet model development

### ISSUE THREAD

**Issue by nburner96** [amaiya/ktrain#489]
I have a script that performs 5-fold cross validation on an image set with the pretrained Resnet50 model. A model for each fold is trained using 30 epochs, and the CV is repeated 5 additional times. 

The first fold of the first run trains fast enough for my purposes, about 6 minutes per epoch:

````
Epoch 1/10
288/288 [==============================] - 370s 1s/step - loss: 13.0123 - mse: 13.0123 - val_loss: 2.8116 - val_mse: 2.8116
Epoch 2/10
288/288 [==============================] - 367s 1s/step - loss: 6.2146 - mse: 6.2146 - val_loss: 2.2179 - val_mse: 2.2179
````

However, the training time for each epoch in successive models is substantially higher, about 30 minutes for a single epoch.

```
begin training using onecycle policy with max lr of 0.0001...
Epoch 1/10
286/286 [==============================] - 2229s 8s/step - loss: 8.5098 - mse: 8.5098 - val_loss: 2.3128 - val_mse: 2.3128
Epoch 2/10
286/286 [==============================] - 2213s 8s/step - loss: 5.2229 - mse: 5.2229 - val_loss: 2.4311 - val_mse: 2.4311
```

At the end of each fold I use the Ktrain function `release_gpu_memory` to attempt to clear the current session so that each iteration "starts fresh". A common solution I see is to use the Keras `clear_session()` function, which this function includes. However, it appears not to be helping. What can I do to keep a consistent training time across each iteration? Below is my script:

```
import os
import ktrain
from ktrain import vision as vis
from ktrain.vision.data import images_from_df
from ktrain.core import release_gpu_memory
import pandas as pd
from pprint import pprint
import glob2
from sklearn.model_selection import (GroupShuffleSplit, StratifiedGroupKFold)
from sklearn.model_selection import GroupKFold
import matplotlib.pyplot as plt
from IPython.display import display, HTML
import numpy as np
import tensorflow as tf
import multiprocessing

#############################
# Initial settings excluded #
#############################

# Intialize results df
results = labels.copy()

for r in range(runs):

    # Generate splits
    gkf = GroupKFold(n_splits=5)
    
    for i, (train_index, val_index) in enumerate(gkf.split(filtered, groups=filtered[group]), 1): # start at fold 1, not 0
        
        # Create training and validation data sets and image generators
        train, val = filtered.iloc[train_index], filtered.iloc[val_index]
        (train_img, val_img, preproc) = images_from_df(train_df=train, image_column='id', label_columns='DIFF',directory=f"images/{expt}", suffix='.tif', 
                                                       val_df=val, is_regression=True, target_size=dim, color_mode='rgb')
        
        # Create model
        model = vis.image_regression_model(name='pretrained_resnet50', train_data=train_img, val_data=val_img, 
                                           freeze_layers=None, metrics=['mse'])

        learner = ktrain.get_learner(model=model, train_data=train_img, val_data=val_img, 
                                     workers=multiprocessing.cpu_count()-1, use_multiprocessing=False, batch_size=64)
        
        # Train model
        print(f'Run {r+1} Fold {i}')
        
        learner.fit_onecycle(1e-4, epochs)
        
        # PLot training and validation loss
        learner.plot()
        
        # Create Predictor instance
        predictor = ktrain.get_predictor(learner.model, preproc)

        def predict_diff(row):
            id = row['id']
            fname = f'images/{expt}/{id}.tif'
            pred = round(predictor.predict_filename(fname, return_proba=True)[0])
            return pred

        mask = results.id.isin(val.id)

        results.loc[mask, f'Run_{r+1}'] = results[mask].apply(lambda row: predict_diff(row), axis=1)
        
        release_gpu_memory()


results.to_csv("21PLTR-NNN_PED.csv", index=False)
```

**Comment by nburner96** [comment]
I was attempting to recreate a reproducible example similar to the Age Prediction example given on this page. In the process of doing so I noticed that subsequent epochs after the first fold exhibited normal training times when predictions are not made in the previous iteration. In the code posted above, when I comment out:

`results.loc[mask, f'Run_{r+1}'] = results[mask].apply(lambda row: predict_diff(row), axis=1)`

...which goes through the data frame column with the filenames and makes predictions, subsequent epochs have normal training times. I thought this might have been an issue with the data frame being retained, so I modified the script to read, make predictions, and write it each iteration, but that did not solve the problem. Therefore, I believe the problem has to do with the predictor function. Any thoughts as to why this could be?

**Comment by amaiya** [comment]
I've called both `predictor.predict` and `predictor.predict_filename` on my end  when trying to reproduce and didn't see any slow down at all (all epochs trained in the same amount of time).  If you can post your example, it would help investigate what is happening.

**Comment by nburner96** [comment]
Okay, I was able to replicate the problem perfectly. For reference, I did this on Jupyter notebook if that makes any difference. I used ~18,000 photos that were used in the Age Prediction demo, which I stored in a folder called "example". I only ran two epochs, the first one ran in around 4 minutes, and the estimated time for the second epoch was around 25 minutes. Here is the script:

```
import os
import ktrain
from ktrain import vision as vis
from ktrain.vision.data import images_from_df
from ktrain.core import release_gpu_memory
import pandas as pd
from pprint import pprint
from sklearn.model_selection import (GroupShuffleSplit, StratifiedGroupKFold, KFold)
from sklearn.model_selection import GroupKFold
import matplotlib.pyplot as plt
from IPython.display import display, HTML
import numpy as np
import tensorflow as tf
import multiprocessing
from keras import backend as K
import glob2

runs = 2
epochs = 1
batch_size = 64

# build a regular expression that extracts the age from file name
PATTERN = r'([^\\]+)_\d+_\d+_\d+.jpg$'
import re
p = re.compile(PATTERN)
r = p.search('example\\40_1_0_20170116222110661.jpg') # the double slash shows up in the path on windows
print("Extracted Age:%s" % (int(r.group(1))))

# kfold
DATADIR='example'

for r in range(runs):
    (train_data, val_data, preproc) = vis.images_from_fname(train_folder=DATADIR, pattern = PATTERN,
                                                        is_regression=True, target_size=(60, 300),
                                                           val_pct=0.3)
    
    

    model = vis.image_regression_model('pretrained_resnet50', train_data, val_data)

    learner = ktrain.get_learner(model=model, train_data=train_data, val_data=val_data, 
                                 workers=8, use_multiprocessing=False, batch_size=batch_size)

    learner.fit_onecycle(1e-4, epochs)

    # PLot training and validation loss
    learner.plot()
    
    predictor = ktrain.get_predictor(learner.model, preproc)
    
    for file in val_data.filenames:
        fname = f'example/{file}'
        pred = round(predictor.predict_filename(fname, verbose=0)[0])
```

**Comment by nburner96** [comment]
Tried this as a script in PyCharm and the same result occurred as with Jupyter, FYI

**Comment by amaiya** [comment]
Hi @nburner96  I ran your code in a Jupyter notebook on Ubuntu 20.04 and was not able to reproduce this problem.  I used the UTKFace dataset [available here](https://drive.google.com/drive/folders/0BxYys69jI14kU0I1YUQyY1ZDRUE?resourcekey=0-01Pth1hq20K4kuGVkp3oBw) and used your code exactly.  The training times shown by Keras were roughly 30 seconds on a Titan V GPU for both runs (i.e., no slowdown occurred for second run):
```
# RUN #2
begin training using onecycle policy with max lr of 0.0001...
260/260 [==============================] - 27s 93ms/step - loss: 389.7512 - mae: 14.7477 - val_loss: 227.5112 - val_mae: 11.5030
```

This suggests that it is either something local to your machine and/or some strange issue in Windows (as it appears you're using Windows as opposed to Linux), which is why I had recommended earlier to try to reproduce the issue on Google Colab. You might try different ways of running predictions (e.g., `predict_folder` instead of `predict_filename`, preprocessing the prediction yourself using Keras, and numpy etc.) to shed more light on what is going on. 

I will close this issue for now, since it does not appear to be an issue in the codebase.   But, feel free to respond to this thread if you have further questions.

**Comment by nburner96** [comment]
It was a issue with Windows. I ran the script on a Linux virtual machine and there were no issues.
